test/e2e/multihoming.go

```diff
@@ -1194,3 +1242,66 @@ func kickstartPod(cs clientset.Interface, configuration podConfiguration) *v1.Po
 	}, 2*time.Minute, 6*time.Second).Should(Equal(v1.PodRunning))
 	return createdPod
 }
+
+func createNad(namespace string, nadClient nadclient.K8sCniCncfIoV1Interface, netConfig networkAttachmentConfig) {
+	By("creating the attachment configuration")
+	_, err := nadClient.NetworkAttachmentDefinitions(namespace).Create(
+		context.Background(),
+		generateNAD(netConfig),
+		metav1.CreateOptions{},
+	)
+	ExpectWithOffset(1, err).NotTo(HaveOccurred())
+}
+
+func setupLocalnetUnderlay(cs clientset.Interface, netConfig networkAttachmentConfig) func() {
+	By("setting up the localnet underlay")
+	nodes := ovsPods(cs)
+	ExpectWithOffset(1, nodes).NotTo(BeEmpty())
+
+	const secondaryInterfaceName = "eth1"
+	ExpectWithOffset(1, setupUnderlay(nodes, secondaryInterfaceName, netConfig)).To(Succeed())
+	return func() {
+		By("tearing down the localnet underlay")
+		ExpectWithOffset(1, teardownUnderlay(nodes)).To(Succeed())
+	}
+}
+
+func addUnderlayBridgeIP(networkName string, ip string, subnet string) func() {
+	By("adding IP to the underlay docker bridge")
+	cli, err := client.NewClientWithOpts(client.FromEnv)
+	ExpectWithOffset(1, err).NotTo(HaveOccurred())
+
+	gatewayIP, err := getNetworkGateway(cli, networkName)
+	ExpectWithOffset(1, err).NotTo(HaveOccurred())
+
+	underlayBridgeName, err := findInterfaceByIP(gatewayIP)
+	ExpectWithOffset(1, err).NotTo(HaveOccurred())
+
+	cmd := exec.Command("sudo", "ip", "addr", "add", ip+subnet, "dev", underlayBridgeName)
+	cmd.Stderr = os.Stderr
+	err = cmd.Run()
+	ExpectWithOffset(1, err).NotTo(HaveOccurred())
+
+	return func() {
+		cmd := exec.Command("sudo", "ip", "addr", "del", ip+subnet, "dev", underlayBridgeName)
+		cmd.Stderr = os.Stderr
+		err = cmd.Run()
+		ExpectWithOffset(1, err).NotTo(HaveOccurred())
+	}
+}
+
+func createUnderlayService(ip string, port string) func() {
```

Comment:
I would rather have the setup section in the context's BeforeEach and the teardown in the AfterEach .

Plus, return errors and assert them in the test's body / setup / teardown. The test becomes IMO more readable.